toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -165,91 +59,176 @@ static auto GetBinaryOp(TokenKind kind) -> Semantics::BinaryOperator::Op {
-    -> Semantics::Statement {
-  CARBON_CHECK(parse_tree().node_kind(node) == ParseNodeKind::ReturnType());
+        auto name_node = node_stack.back().parse_node;
+        sem_ir.push_back(semantics_.nodes_.Store(Semantics::SetName(
+            name_node, parse_tree().GetNodeText(name_node), function_id)));
+        node_stack.pop_back();
 
-  auto subtree = ParseSubtreeConsumer::ForParent(parse_tree(), node);
-  return TransformExpression(subtree.RequireConsume());
+        // Do nothing with the `fn`.
+        node_stack.pop_back();
+
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir)));
+        break;
+      }
+      case ParseNodeKind::FileEnd(): {
+        ++it;
+        CARBON_CHECK(it == range.end())
+            << "FileEnd should always be last, found "
+            << parse_tree().node_kind(*it);
+
+        for (const auto& entry : node_stack) {
+          semantics_.root_block_.append(entry.sem_ir.begin(),
+                                        entry.sem_ir.end());
+        }
+        return;
+      }
+      case ParseNodeKind::InfixOperator(): {
+        llvm::SmallVector<Semantics::NodeRef> sem_ir;
+
+        sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                      node_stack.back().sem_ir.end());
+        auto rhs_id = *node_stack.back().result_id;
+        node_stack.pop_back();
+
+        sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                      node_stack.back().sem_ir.end());
+        auto lhs_id = *node_stack.back().result_id;
+        node_stack.pop_back();
+
+        // Figure out the operator for the token.
+        auto token = parse_tree().node_token(parse_node);
+        auto token_kind = tokens_->GetKind(token);
+        auto op = GetBinaryOp(token_kind);
+
+        auto literal_id = next_id();
+        sem_ir.push_back(semantics_.nodes_.Store(Semantics::BinaryOperator(
+            parse_node, literal_id, op, lhs_id, rhs_id)));
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir), literal_id));
+        break;
+      }
+      case ParseNodeKind::Literal(): {
+        auto literal_id = next_id();
+
+        llvm::SmallVector<Semantics::NodeRef> sem_ir;
+        auto token = parse_tree().node_token(parse_node);
+        switch (auto token_kind = tokens_->GetKind(token)) {
+          case TokenKind::IntegerLiteral(): {
+            sem_ir.push_back(semantics_.nodes_.Store(Semantics::IntegerLiteral(
+                parse_node, literal_id, tokens_->GetIntegerLiteral(token))));
+            break;
+          }
+          default:
+            CARBON_FATAL() << "Unhandled kind: " << token_kind.Name();
+        }
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir), literal_id));
+        break;
+      }
+      case ParseNodeKind::ReturnStatement(): {
+        // TODO: For now just blindly pop what should be StatementEnd.
```

Comment:
Can we `CARBON_CHECK` that it has the expected parse node kind?

Reply:
Done, but for future reference, all the TODOs about ReturnStatement mean that this change won't last long at all. the `;` shouldn't be a child.